src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal class CircuitRegistry
+    {
+        private readonly ComponentsServerOptions _options;
+        private readonly ILogger _logger;
+        private readonly PostEvictionCallbackRegistration _postEvictionCallback;
+
+        public CircuitRegistry(
+            IOptions<ComponentsServerOptions> options,
+            ILogger<CircuitRegistry> logger)
+        {
+            _options = options.Value;
+            _logger = logger;
+
+            ActiveCircuits = new ConcurrentDictionary<string, CircuitHost>(StringComparer.Ordinal);
+
+            InactiveCircuits = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _options.MaxRetainedDisconnectedCircuits,
+            });
+
+            _postEvictionCallback = new PostEvictionCallbackRegistration
+            {
+                EvictionCallback = OnEntryEvicted,
+            };
+        }
+
+        internal ConcurrentDictionary<string, CircuitHost> ActiveCircuits { get; }
+
+        internal MemoryCache InactiveCircuits { get; }
+
+        public void Register(CircuitHost circuitHost)
+        {
+            ActiveCircuits.TryAdd(circuitHost.CircuitId, circuitHost);
+        }
+
+        public void Deactivate(CircuitHost circuitHost, string connectionId)
+        {
+            if (!ActiveCircuits.TryGetValue(circuitHost.CircuitId, out circuitHost))
+            {
+                // The circuit might already have been marked as inactive.
+                return;
+            }
+
+            if (!string.Equals(circuitHost.CircuitClient.ConnectionId, connectionId, StringComparison.Ordinal))
+            {
+                // The circuit is associated with a different connection. One way this could happen is when
+                // the client reconnects with a new connection before the OnDisconnect for the older
+                // connection is executed. Do nothing
+                return;
+            }
+
+            ActiveCircuits.TryRemove(circuitHost.CircuitId, out circuitHost);
```

Comment:
So operations like this between like 58 and here just **fundamentally** have a race condition. Someone else could *claim* the circuit between these two lines of code. Or they could claim it between here and line 68.

Reply:
We should also probably look at the result of `TryRemove`

Reply:
I'll move the locking in to the registry and ignore these comments.